Added single course page

ellaml opened the pull request:
- Added course page to `urls.py` that uses `course_id `as parameter in the url, and
  loads `course.html` template. When the `course_id `is not found it
  redirects to Courses page.
- Added redirection from Courses page to a specific course when clicking on its row in the table.
- Separated `single-review-box` element to a new `single_review.html`
  template in order to include it easily in other files
- Updated Reviews page to Single Course page to include
  `single_review.html` template with review as passed variable
- Added tests for valid course in url using our courses test data, and for redirection in case of invalid course

The logic for the buttons "Syllabi", "Add Review", "Follow Course" will be added later on. (They depend on other PRs)

fix: #95

ellaml committed:
Added single course page

- Added course page to "urls.py" that uses course_id as parameter in the url, and
  loads "course.html" template. When the course_id is not found it
  redirects to Courses page.
- Seperated single-review-box element to a new "single_review.html"
  template in order to include it easily in other files
- Updated Reviews page to Single Course page to include
  "single_review.html" template with review as passed variable
- Added tests for valid course in url using our courses test data, and for redirection in case of invalid course

ellaml commented:
https://user-images.githubusercontent.com/73358484/117580310-0c55d480-b100-11eb-97bb-3791f9be8a76.mp4

oriAdler commented on homepage/templates/homepage/courses/course.html:
```diff
@@ -0,0 +1,75 @@
+        </div>
+    </div>
+    <div class="bottom-container">
+        <div class="num-review-title">{{reviews.count}} reviews available</div>
```
```suggestion
        {% if reviews|length > 0 %}
        <div class="num-review-title">{{reviews.count}} reviews available</div>
        ...
        {% else %}
        <div class="num-review-title">no reviews available</div>
        {% endif %}
```
(optional)

ellaml replied:
@oriAdler I'll leave it that way for now - since we need to update Reviews page as well.
We can do it as part of our "beautifying" PR next week

oriAdler commented on homepage/tests/course_tests.py:
```diff
@@ -0,0 +1,24 @@
+    ])
+@pytest.mark.django_db
+def test_renders_course_template(client, valid_course_id):
+    response = client.get('/course/' + str(valid_course_id) + '/')
```
```suggestion
    response = client.get(f'/course/{valid_course_id}/')
```

ellaml replied:
@oriAdler I tried to do it this way but it fails the test - the variables are not strings

oriAdler replied:
@ellaml my bad, the last ```/``` is redundant.
It supposed to work now. There are similar tests on PR #106

oriAdler replied:
@ellaml There are similar tests on PR #106, also with f-Strings and client maybe it can help.
Anyway, if it doesn't work, you can leave it that way.

ellaml replied:
@oriAdler Fixed. 
My bad, I had another problem.

oriAdler commented on homepage/tests/course_tests.py:
```diff
@@ -0,0 +1,24 @@
+def test_invalid_course_resquest(client):
+    invalid_id = 11111
+
+    response = client.get('/course/' + str(invalid_id) + '/')
```
```suggestion
    response = client.get(f'/course/{invalid_id}/')

```

ellaml replied:
@oriAdler  I tried to do it this way but it fails the test - the variables are not strings

ellaml replied:
@oriAdler  same ^^

oriAdler commented on homepage/views.py:
```diff
@@ -53,3 +54,12 @@ def add_review(request):
+def course(request, id):
+    try:
+        course = Course.objects.get(pk=id)
+        reviews = Review.objects.filter(course=id)
```
```suggestion
        reviews = Review.objects.filter(course=id).order_by('-avg_rating')
```
WDYT?

VeraTs replied:
imo reviews should be deisplayed according to date

LeonRabinovich replied:
IMO reviews should be displayed by the number of likes as a default 
(maybe add a button to change the display by date / ratings)
Meaning the review that has the most likes will be displayed first.

I think its more helpful than displaying by avg-rating or by date, since as a user i will usually want to look at the "best" reviews first and not the newest

ellaml replied:
Decided to sort by likes as Leon suggested

oriAdler reviewed (commented):
Looks Awesome as always:)
I've left a few comments

VeraTs reviewed (commented):
Looks great, I agree with Ori's comments, basically

VeraTs commented on homepage/templates/homepage/courses/course.html:
```diff
@@ -0,0 +1,75 @@
+                    <label class="raters">({{course.num_of_raters}} Raters)</label>
+                </div>
+                <div class="credit-container">
+                    <label>Credit Points:</label>
```
suggestion: credit points to switch places with course load in the display, so that course load and rating are on the same level?

ellaml replied:
switched

LeonRabinovich commented on homepage/tests/course_tests.py:
```diff
@@ -0,0 +1,24 @@
+@pytest.mark.django_db
+def test_renders_course_template(client, valid_course_id):
+    response = client.get('/course/' + str(valid_course_id) + '/')
+
```
Kobi suggested in previous reviews that we remove the blank lines before the 'assert', so i think it needs to be removed here as well

ellaml replied:
Missed that one, if that's important in Kobi's opinion I'll update of course.
I like spaces for readability anyway :)

ellaml replied:
I had to make another update so I fixed it on the way :)

LeonRabinovich commented:
Looks great,
I agree with Ori's review as well and added some things

ellaml committed:
UI changes and addition of sorting

- Swapped course load and credit points location
- Ordered reviews by num of likes
- Small code fixes

ellaml committed:
resolved conflicts with main

LeonRabinovich commented:
LGTM

ellaml committed:
fix course_tests formatting

oriAdler reviewed (approved):
LGTM

kobihk commented on homepage/templates/homepage/reviews/single_review.html:
```diff
@@ -0,0 +1,64 @@
+        About Course: {{review.course.name}}
+    </div>
+    <div style="margin-top: 20px">
+        {% if review.rate == 5 %}
+        {% for i in "12345"|make_list %}
+        <img src="{% static 'icons/star.png' %}">
+        {% endfor %}
+        {% elif review.rate == 4 %}
+        {% for i in "1234"|make_list %}
+        <img src="{% static 'icons/star.png' %}">
+        {% endfor %}
+        {% elif review.rate == 3 %}
+        {% for i in "123"|make_list %}
+        <img src="{% static 'icons/star.png' %}">
+        {% endfor %}
+        {% elif review.rate == 2 %}
+        {% for i in "12"|make_list %}
+        <img src="{% static 'icons/star.png' %}">
+        {% endfor %}
+        {% else %}
+        <img src="{% static 'icons/star.png' %}">
+        {% endif %}
```
maybe you can do it simpler something like:
https://stackoverflow.com/a/17691875
```suggestion
        {% for i in range(review.rate) %}
        <img src="{% static 'icons/star.png' %}">
        {% endfor %}
```

ellaml replied:
@kobihk I tried it but I get a template syntax error.
According to this discussion https://stackoverflow.com/questions/1107737/numeric-for-loop-in-django-templates
and this bug report https://code.djangoproject.com/ticket/5172, it looks like Django does not support it. The link you've attached is referring to Jinja.

kobihk replied:
@ellaml please try with this code:
```suggestion
        {% for i in " "|rjust:review.rate|make_list %}
        <img src="{% static 'icons/star.png' %}">
        {% endfor %}

```

ellaml committed:
Fixed for loop in single review template
